### ISSUE TITLE
Hide Java.type(...) and Bean(...) instructions

### ISSUE THREAD

**Issue by madoar** [PhoenicisOrg/scripts#1070]
I discovered that it is problematic to add `Java.type` and `Bean` instructions to the global scope at the top of the script files, because this can lead to "already defined" errors if multiple scripts define the "same" value.

To prevent this from happening I propose to hide the `Java.type` and `Bean` instructions between an included script. Where every `Java.type` and `Bean` instruction is always defined in a single script.

For more details see https://github.com/PhoenicisOrg/scripts/pull/1066#discussion_r300856861

**Comment by plata** [comment]
Can we do this automatically for Beans?

**Comment by madoar** [comment]
What do you mean by automatically?

**Comment by plata** [comment]
I wouldn't want to manually write a script for every Bean which hides the Bean in an `include`.

**Comment by madoar** [comment]
I don't think this is feasible. We could write an application that creates a script for a given `@Safe` annotated class, in my opinion the problem is that writing such an application requires a lot more effort than writing the scripts ourselves by hand.

**Comment by qparis** [comment]
I don't see the problem with the current Bean implementation..

**Comment by madoar** [comment]
The problem is not our `Bean` implementation. It is about how we access the `Bean` objects.
I don't like if we use the `Bean` objects inline, e.g. like: 

```javascript
const value = Bean("safeBean").getX().fetchValue();
```

This style is really ugly if you have multiple such lines in a script. Therefore I tried to move the `Bean` declaration outside of the methods/classes in the script to the header part:

```javascript
include("a");
include("b");
...
const safeBean = Bean("safeBean");
...
...
const value = safeBean.getX().fetchValue();
...
```

This style has then the problem that if any of the imported script `a` or `b` (or one of their imported scripts) also define `const safeBean` we get a nasty JavaScript error because `safeBean` has already been defined.

We can circumvent this problem if we "hide" our Bean usage behind `include` commands and we require that every Bean is used in at most one included script.

**Comment by madoar** [comment]
This also has the additional benefit that we can much easier do changes to our `Bean` implementations because we just need to change a single JavaScript file (the one that uses the `Bean` directly).

**Comment by qparis** [comment]
I see. What about scoping all modules inside a function?

**Comment by qparis** [comment]
What I am thinking about is really close to ES imports. I think we can easily implement them

**Comment by madoar** [comment]
What do you mean by modules?

**Comment by madoar** [comment]
Can you give an example? Please also see https://github.com/PhoenicisOrg/phoenicis/issues/2042

**Comment by qparis** [comment]
Let say that you have include("a") with const bean = Bean("bean") inside. 

`include("a")
const bean = Bean("bean")`

raises an error, right? 

Now, imagine include("a") is resolved with function() { const bean = Bean("bean") }(

No problem anymore. This is more or less what ES modules are doing

**Comment by madoar** [comment]
Yes, but we normally want to expose some parts of `include("a")` to the calling script, right? How would we achieve this if we evaluate the content of `"a"` inside its own scope?

**Comment by qparis** [comment]
This is where exports are required

**Comment by qparis** [comment]
https://www.jvandemo.com/a-10-minute-primer-to-javascript-modules-module-formats-module-loaders-and-module-bundlers/

**Comment by plata** [comment]
Couldn't we then do?:
```Javascript
const myBean = include("a")
```
where
```Javascript
function() { return Bean("bean") }()
```

**Comment by qparis** [comment]
Yes, the idea is to return dictionnary so that we can export multiple objects

**Comment by plata** [comment]
Could you prototype your idea?

**Comment by madoar** [comment]
I'm not sure if it is a good idea to re-implement the whole ECMAScript Module system when we could instead try to use the implementation provided by graal.js? I think we should continue the current discussion in PhoenicisOrg/phoenicis#2042 which is exactly about this question.

Still I don't think that improving or changing our import mechanism solves this issue. I still believe that it is a good idea to have a single location where we use each `Bean`. We already use this approach with the functionality of the `fileUtilities` Bean, where the only location it is directly accessed is the https://github.com/PhoenicisOrg/scripts/blob/master/Utils/Functions/Filesystem/Files/script.js script.

Such a single access location approach makes it much easier to do changes in the Bean interface on the Java side, without the need to perform larger changes in the scripts.

**Comment by qparis** [comment]
The problem we are talking about is not just about bean in fact... It will happen everytime we have two imports that use a same variable name

**Comment by madoar** [comment]
That is true. Still I don't think this should be part of this issue. In my opinion it should be part of either a new issue or of PhoenicisOrg/phoenicis#2042.

**Comment by qparis** [comment]
Yes. However, considering this, I think hiding Bean instruction is less urgent

**Comment by madoar** [comment]
Why?

In my opinion a `Bean("xyz").something()` looks out of place. It opens multiple questions to the developer:
- what does `Bean(...)` do at all? (it is not part of standard JavaScript)
- how is the interface of `xyz` or `Bean("xyz")` defined?
- the method `something()` is typically defined in Java so do we need some additional conversions to use the function in a natural manner in JavaScript? (e.g. `Java.from` and `Java.to`)

All these questions can be solved/prevented by hiding the direct `Bean` usage behind an `import`/`include` because `import`/`include` is (or at least feels like) native JavaScript.

**Comment by madoar** [comment]
The earlier we reach a finalization of the script API the better it is for the community in my opinion, because they don't need to change their habits later on if we change/extend an unfinished API.

**Comment by plata** [comment]
> In my opinion a Bean("xyz").something() looks out of place

Actually, we already hide the usage of `Beans` already (in QuickScripts, plugins, settings etc.). So in the vast majority of cases, you will not run into `Beans` when developing an application script. Therefore, I don't think that it affects the community a lot.

**Comment by madoar** [comment]
I agree it shouldn't affect the community a lot, it is just a more fundamental decision. 

If we take https://github.com/PhoenicisOrg/phoenicis/issues/2042 in scope we could also think whether we require the `Bean` syntax at all. Instead we could "hide" the Beans behind JS proxy classes that forward the calls to the Java (Bean) objects. This would require a custom loader that supports this of course.

**Comment by plata** [comment]
If we want to make the scripts completely agnostic like @qparis suggested, that will not be enough (at least for `Java.type`). Apparently, `Bean` does not work like intended (for the reasons you've described). But can't this be fixed in https://github.com/PhoenicisOrg/phoenicis/blob/240e682b6529f1a134fe8e3ecb9453e24a99b458/phoenicis-scripts/src/main/java/org/phoenicis/scripts/engine/injectors/BeanInjector.java similarly to https://github.com/PhoenicisOrg/phoenicis/blob/240e682b6529f1a134fe8e3ecb9453e24a99b458/phoenicis-scripts/src/main/java/org/phoenicis/scripts/engine/injectors/IncludeInjector.java (i.e. by storing the already injected Beans to avoid "already defined")?

**Comment by madoar** [comment]
You're right. Still this alone won't fix the issue with reassignments of beans, i.e.:

```javascript
const bean = Bean(...);
...
const bean = Bean(...);
```

**Comment by plata** [comment]
That's true. We would need something like module scope for variables.

**Comment by madoar** [comment]
I believe this to be fixed by #1076. Now every script can redefine their own `bean` variable without any conflicts.
